### PR TITLE
New Data Structure for Source Lines to replace SDB

### LINKED ISSUE #843: Refactor RzBinFile.sdb_addrinfo out of Sdb
https://github.com/rizinorg/rizin/blob/31a2406f461d8f15c5789b855e7eaa9cb4dc120e/librz/include/rz_bin.h#L297

This is an sdb with records like `0x424242=some/filename.c:1337`, next to some other entries where the dwarf code abuses this sdb for completely different purposes.
This could be done much better in either a hashtable or even better a tree (so queries like "closest addr info" are possible). Also the filenames should be in a string pool to reduce bloat.

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**DO NOT SQUASH**

Fix #843

Overview:
* Both `RzBinSourceRow` and `sdb_addrinfo` are removed. Instead:
* Added a new data structure `RzBinSourceLineInfo` for holding a large amount of addr+line+column+filename samples in a memory-efficient way with O(log n) access by address.
* An `RzBinSourceLineInfo` can be conveniently constructed using the `RzBinSourceLineInfoBuilder` by pushing samples in an arbitrary order into it and the builder will take care of creating a valid `RzBinSourceLineInfo` in a final step.
* Such an `RzBinSourceLineInfo` object is part of `RzBinObject` next to symbols, sections, etc. and can be in a bin plugin's `lines` function, as it is done for dex and CoreSymbolication.
* `CL` which wasn't really using meta but just messing around with the `sdb_addrinfo` has been removed.
* `ix` has been added (`iX` removed and integrated here):
    ```
    Usage: ix   Display source file line info (from debug info)
    | ix[j]   List all source line info available (this was CL before)
    | ix.[j]  Show source line info for current address (this was CL. before)
    | ixf[j]  Show summary of all source files used (this was iX before)
    ```

For #676, this reduces the load time for me from ~1m to ~45s.
Memory usage (from massif) after loading this binary:
Before:
20210325_21h14m23s_grim
After:
20210325_21h13m48s_grim
So we get ~180MB vs. ~540MB. Getting lower than this is probably impossible while maintaining efficient access.

### CODE REVIEW

**Comment by XVilka** [comment]
Would be nice to reuse it also for PDB too.

**Comment by thestr4ng3r** [comment]
Yes, it should be used for everything eventually. But actually I am not even sure if pdb right now supports line info at all.

File: librz/bin/dwarf.c [hunk 1]
```diff
@@ -807,37 +835,33 @@ RZ_API void rz_bin_dwarf_line_header_reset_regs(const RzBinDwarfLineHeader *hdr,
...
-	rz_list_push(rows_out, row);
+	rz_bin_source_line_info_builder_push_sample(bob, regs->address, line, (ut32)regs->column, file);
 }
 
+/**
+ * \brief Execute a single line op on regs and optionally store the resulting line info in bob
+ */
 RZ_API bool rz_bin_dwarf_line_op_run(const RzBinDwarfLineHeader *hdr, RzBinDwarfSMRegisters *regs, RzBinDwarfLineOp *op,
-	RZ_NULLABLE RZ_OUT RzList /*<RzBinSourceRow>*/ *rows_out, RZ_NULLABLE RzBinDwarfDebugInfo *info) {
+	RZ_NULLABLE RzBinSourceLineInfoBuilder *bob, RZ_NULLABLE RzBinDwarfDebugInfo *info, RZ_NULLABLE RzBinDwarfLineFileCache fnc) {
```
Comment (XVilka) [inline librz/bin/dwarf.c:851]: Why Bob and not Alice?

Comment (thestr4ng3r) [reply librz/bin/dwarf.c:851]: Bob the builder is the builder

Comment (thestr4ng3r) [reply librz/bin/dwarf.c:851]: Added one alice case

File: librz/bin/bfile.c [hunk 2]
```diff
@@ -700,10 +698,6 @@ RZ_API void rz_bin_file_free(void /*RzBinFile*/ *_bf) {
 		bf->curxtr->free_xtr((void *)(bf->xtr_obj));
 	}
 	// TODO: unset related sdb namespaces
```
Comment (wargio) [inline librz/bin/bfile.c]: useless comment?

**Review by XVilka — APPROVED** [review]
I went through and it's a good quality code! 👍 
While I think the API should be more abstracted from the fact that we use DWARF, PDB, or dlSYM, it's already a huge improvement and we can merge it right away, imho. Further improvements could be done as a separate issues/PRs

**Comment by XVilka** [comment]
I tried this branch on rizin itself build with debug information (`rizin $(which rizin)` then `aaa; s main; Vp`):

image

These `rizin.c34rizin.c` strings do not look very good. Could you please address this problem too?

File: librz/bin/dbginfo.c [hunk 3]
```diff
@@ -1,130 +1,261 @@
...
+	rz_str_constpool_init(&builder->filename_pool);
+}
+
+RZ_API void rz_bin_source_line_info_builder_fini(RzBinSourceLineInfoBuilder *builder) {
+	rz_vector_fini(&builder->samples);
+	rz_str_constpool_fini(&builder->filename_pool);
+}
+
+/**
+ * \param line may be 0 or a positive line number, see RzBinSourceLine for exact meaning
```
Comment (ret2libc) [inline librz/bin/dbginfo.c]: a description for this function please?

**Comment by thestr4ng3r** [comment]
> These `rizin.c34rizin.c` strings do not look very good. Could you please address this problem too?

Yes. The main problem here is that we don't have any tests for this (yet) as it depends on your local filesystem contents.

**Comment by thestr4ng3r** [comment]
> While I think the API should be more abstracted from the fact that we use DWARF, PDB, or dlSYM,...

Added detailed info about going more format-agnostic here: https://github.com/rizinorg/rizin/issues/907#issuecomment-808046871

**Comment by thestr4ng3r** [comment]
> line - An unsigned integer indicating a source line number. Lines are numbered beginning at 1. The compiler may emit the value 0 in cases where an instruction cannot be attributed to any source line. 

This isn't handled nicely yet, I have to fix it.

**Comment by thestr4ng3r** [comment]
All fixed, including the ugly `rizin.c34rizin.c` comments.

**Comment by XVilka** [comment]
Two failing tests:
```


[XX] db/formats/mach0/coresymbolication Xcode symbols cache new (line numbers)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'ix
?e --
ix.@`is~viewDidLoad[2]`
ix.@0x100006430
' bins/mach0/D9A37B67-F10A-35AA-A852-ABFBBECC03AE-new.symbols
-- stdout
@@ -5,7 +5,6 @@
 0x100006404	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	27
 0x100006420	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
 0x100006430	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
-0x100006440	-	-
 0x100006440	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	-
 0x100006444	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
 0x100006454	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
@@ -30,6 +29,7 @@
 0x100006528	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.m	51
 0x10000652c	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	13
 0x10000653c	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	13
+0x100006540	-	-
 0x100006540	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	-
 0x100006550	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.m	15
 0x100006564	-	-




[XX] db/formats/mach0/coresymbolication Xcode symbols cache (line numbers)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'ix
?e --
ix.@`is~viewDidLoad[2]`
ix.@0x100006430
' bins/mach0/D9A37B67-F10A-35AA-A852-ABFBBECC03AE.symbols
-- stdout
@@ -5,7 +5,6 @@
 0x100006404	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	27
 0x100006420	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
 0x100006430	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
-0x100006440	-	-
 0x100006440	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	-
 0x100006444	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
 0x100006454	/Users/ftamagni/src/TestRTTI/TestRTTI/AppDelegate.m	30
@@ -30,6 +29,7 @@
 0x100006528	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.m	55
 0x10000652c	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	13
 0x10000653c	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	13
+0x100006540	-	-
 0x100006540	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.h	-
 0x100006550	/Users/ftamagni/src/TestRTTI/TestRTTI/SceneDelegate.m	15
 0x100006564	-	-

```

**Comment by thestr4ng3r** [comment]
> Two failing tests:

Fixed and updated unit tests to catch this.

**Review by XVilka — APPROVED** [review]

File: librz/include/rz_bin_dwarf.h [hunk 4]
```diff
@@ -939,14 +947,18 @@ RZ_API void rz_bin_dwarf_loc_free(HtUP /*<offset, RzBinDwarfLocList*>*/ *loc_tab
 RZ_API void rz_bin_dwarf_debug_info_free(RzBinDwarfDebugInfo *inf);
 RZ_API void rz_bin_dwarf_debug_abbrev_free(RzBinDwarfDebugAbbrev *da);
 
-RZ_API RzList *rz_bin_dwarf_parse_line(RzBinFile *binfile, RZ_NULLABLE RzBinDwarfDebugInfo *info, RzBinDwarfLineInfoMask mask);
-RZ_API char *rz_bin_dwarf_line_header_get_full_file_path(RZ_NULLABLE RzBinDwarfDebugInfo *info, const RzBinDwarfLineHeader *header, ut64 file_index);
+typedef char **RzBinDwarfLineFileCache;
```
Comment (ret2libc) [inline librz/include/rz_bin_dwarf.h:950]: what does this represent? A small doxygen-doc to explain it?

**Review by ret2libc — APPROVED** [review]
Couldn't review everything :P I hope it won't break anything. I couldn't see the code coverage for this branch on codecov though. It would help to see which part are not covered at all by tests.

File: librz/include/rz_bin_dwarf.h (hunk 4 above)
Comment (XVilka) [reply librz/include/rz_bin_dwarf.h:950]: Lets merge this and test a bit before the release, not to be untested. I think @thestr4ng3r could add a comment in a separate commit.

**Pull request merged**